core/src/main/scala/akka/persistence/cassandra/query/AllPersistenceIdsStage.scala

```diff
@@ -0,0 +1,135 @@
+  */
+@InternalApi private[akka] class AllPersistenceIdsStage(
+    refreshInterval: Option[FiniteDuration],
+    fetchSize: Int,
+    preparedStatement: PreparedStatement,
+    session: Session)
+    extends GraphStage[SourceShape[String]] {
+
+  import AllPersistenceIdsStage._
+
+  val out: Outlet[String] = Outlet("AllPersistenceIds.out")
+
+  val shape: SourceShape[String] = SourceShape(out)
+
+  def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
+    new TimerGraphStageLogicWithLogging(shape) {
+
+      implicit def executionContext: ExecutionContextExecutor =
+        materializer.executionContext
+
+      private var queryInProgress = false
+      private var knownPersistenceIds: Set[String] = Set.empty
+      private var maybeResultSet = Option.empty[ResultSet]
+      private var buffer: Queue[String] = Queue.empty[String]
+
+      private val queryCallback: AsyncCallback[ResultSet] =
+        getAsyncCallback[ResultSet] { rs =>
+          queryInProgress = false
+          maybeResultSet = Some(rs)
+          val available = rs.getAvailableWithoutFetching
+          for (_ <- 1 to available) {
+            val s = rs.one().getString("persistence_id")
+            if (!knownPersistenceIds.contains(s)) {
+              buffer = buffer.enqueue(s)
+              knownPersistenceIds += s
+            }
+          }
+          tryPushOne()
+          if (refreshInterval.isEmpty && buffer.isEmpty && rs.isExhausted) {
+            complete(out)
+          }
+        }
+
+      private def query(): Unit =
+        if (!queryInProgress) {
+          queryInProgress = true
+          val boundStatement = preparedStatement.bind()
+          boundStatement.setFetchSize(fetchSize)
+          session
+            .executeAsync(boundStatement)
+            .asScala
+            .foreach(queryCallback.invoke)
+        }
+
+      private def tryPushOne(): Unit = {
+        while (buffer.nonEmpty && isAvailable(out)) {
+          val (s, newBuffer) = buffer.dequeue
+          buffer = newBuffer
+          push(out, s)
+        }
+      }
+
+      override def preStart(): Unit = {
+        query()
+        refreshInterval.foreach { interval =>
+          val initial =
+            if (interval >= 2.seconds)
+              (interval / 2) + ThreadLocalRandom.current().nextLong(interval.toMillis / 2).millis
+            else interval
+          schedulePeriodicallyWithInitialDelay(Continue, initial, interval)
+        }
+      }
+
+      override def onTimer(timerKey: Any): Unit = {
+        timerKey match {
+          case Continue =>
+            query()
+
+          case _ =>
+        }
+      }
+
+      setHandler(
+        out,
+        new OutHandler {
+          def onPull(): Unit = {
+            tryPushOne()
+            if (buffer.isEmpty && isAvailable(out)) {
+              maybeResultSet match {
+                case None =>
+                  query()
+
+                case Some(rs) =>
+                  if (rs.isExhausted) {
```

Comment:
if it is exhausted we fetch more results, doesn't seem right

Reply:
Oops, I turned the criterion around...